token/transfer-hook/interface/src/onchain.rs

```diff
@@ -92,3 +96,514 @@ pub fn add_cpi_accounts_for_execute<'a>(
         .push(AccountMeta::new_readonly(*program_id, false));
     Ok(())
 }
+
+/// Helper to add accounts required for an `ExecuteInstruction` on-chain,
+/// looking through the additional account infos to add the proper accounts.
+///
+/// Note this helper is designed to add the extra accounts that will be
+/// required for a CPI to a transfer hook program. However, the instruction
+/// being provided to this helper is for the program that will CPI to the
+/// transfer hook program. Because of this, we must resolve the extra accounts
+/// for the `ExecuteInstruction` CPI, then add those extra resolved accounts to
+/// the provided instruction.
+#[allow(clippy::too_many_arguments)]
+pub fn add_extra_accounts_for_execute_cpi<'a>(
+    cpi_instruction: &mut Instruction,
+    cpi_account_infos: &mut Vec<AccountInfo<'a>>,
+    program_id: &Pubkey,
+    source_pubkey: &Pubkey,
+    mint_pubkey: &Pubkey,
+    destination_pubkey: &Pubkey,
+    authority_pubkey: &Pubkey,
```

Comment:
How about passing these in as `AccountInfo`s rather than their pubkeys? Then there's no need for the extra extraction later

Reply:
I think that would cause another clone in Token2022's helper.

Reply:
Ehh, maybe not. Let me check.

Reply:
Hmmmm, seems like it might reduce cloning in the SPL Transfer Hook helper, but adds a clone to the mint data in Token2022's helper (`to_vec()` and `StateWithExtensionsOwned`):

```rust
let mint_data = mint_info.try_borrow_data()?.to_vec();
let mint = StateWithExtensionsOwned::<Mint>::unpack(mint_data)?;
if let Some(program_id) = transfer_hook::get_program_id(&mint) {
    add_extra_accounts_for_execute_cpi(
        &mut cpi_instruction,
        &mut cpi_account_infos,
        &program_id,
        source_info,
        mint_info,
        destination_info,
        authority_info,
        amount,
        additional_accounts,
    )?;
}
```

So we shed four clones in the SPL Transfer Hook helper but gain one in the Token2022 helper. Depends on whether a mint data could be so large it's more expensive to clone than four `AccountInfo`s. Seems like it's OK to do.

Reply:
Scratch that. Seems like we can get away with an extra clone on the mint info. I'm going to add this, since I think it nets us at less clones overall.